### ISSUE TITLE
Architectural discussion

### ISSUE THREAD

**Issue by eirrgang** [radical-collaboration/SCALE-MS#6]
See "Specific comments" at issue #2 

Notes
* https://docs.google.com/document/d/1epsZw6PwgKFc_5UAE8gLRbIiwvykxrAe7H4SRq2hm5U/edit
* "We need to contextualize the discussion about (de)serialization within an architectural view of the execution of a dynamic workflow."

**Comment by eirrgang** [comment]
I am not sure the google sketch diagram is helping me.

If we want a packaging diagram and/or deployment diagram, a starting point might be [this translation](http://www.plantuml.com/plantuml/uml/VP3FQiCm3CRlVWgJDqZs0YtTPPSnHc5duIYfYIqR_w6CqdUVTKXGMGW99Fv-iZvzEkK2gLJlL49JW6Q7-k8iKj4F5S4iCAC-n-YiWM83QSXyx0_Oe3RuK2Ru68ZRdRy-nAdx2McPNBXhfDaITkgvUqqKUg06KoVXDsOCDHdSfPVWVQLrkgtYsHuHjxj2ch7eX9bROUvU1tY70ZgNVzsla1UTpGt7wlPGT61O5rQVOUHqGk5qc5uDruj2Sb39t9h8EvxQFoTQHJKbyBZ9igG_OCC1Usih4lwh4vVSsyIRSuEs_6hEouo__PuuA6xOtXFIMBtx0m00).

If we want to identify the roles participating in use cases, [this translation](http://www.plantuml.com/plantuml/uml/RO-nQWCn38PtFuN86oIZfvHqwL9A2NGMFkLwH9OFsIP3wBjNlkQ8EGeUz7-8tv-EkQ2M6jafXO8oCi7kn1D90QMSgWRAExWx_pBFF0KiKn8FcC7o_bzwCFhnvft_IdgzSBe1oJW9zIIYu4Xgw9CoBRzqx9HYhFAeVNBtuthBC-f8MyjNB0YvF8hEcA-GmpSDbPzWPWoM-hBtD13SUz9KILlbkkuPgrYNBeTUR6vwAtH-9TuyDyitArl1dyI2sx3Vv8ETsTwHPAYHVm40) might be a template.

**Comment by mturilli** [comment]
Happy to go with what helps you the most. Your deployment diagram seems reasonable to me. This is an high-level representation meant for kick starting our discussion. Based on that diagram, 

## Target resource
We can offer an implementation for the 'data stager', 'task scheduler', 'task placer' and 'launcher'. These components would be able to launch a executable-based task on a node with the requirement amount and type of resources. For function-based tasks, we would launch a worker that, in turn, would be able to schedule, place and launch functions/methods.

## Client side
We can offer an API to acquire and manage resources, and one for bulk schedule executable-based tasks/workers on the target resources.

**Comment by eirrgang** [comment]
> For function-based tasks, we would launch a worker that, in turn, would be able to schedule, place and launch functions/methods.

I think this is the primary use case, and I think that for simplicity and flexibility, we can (at least initially) implement new functionality in a single worker before generalizing or fully specifying things like the data persistence we've been talking about.

Is this worker for function-based tasks something that is already implemented?

**Comment by mturilli** [comment]
> I think this is the primary use case,

We do not have use case documents so I am not sure what is/are the project's primary use case(s). Speaking with all the teams involved (Boulder, LSU and Virginia) it is clear we have to support both function- and executable-based tasks. Based on our previous technical meeting and our follow up discussions, it is also clear that Virginia is mostly interested in function-based tasks and I am fine with that.

> I think that for simplicity and flexibility, we can (at least initially) implement new functionality in a single worker before generalizing or fully specifying things like the data persistence we've been talking about.

I agree, I would be very happy with an incremental approach to implementing the worker and its functionalities. Would you initially want to focus on executing an 'hello world' function? If so, we would be interested in  discussing how the function and its context should be passed from the user-facing 'client' to the worker. If you agree, should we start from Section 3 of your document, 'Serializable work'? @andre-merzky, would you agree with that?

> Is this worker for function-based tasks something that is already implemented?

I don't think so. At the moment, it is not clear to me what is the relationship among the worker,  gmxapi and gromacs. Without that clarity, I am not able to start to envision a design for the worker. Also, it is not clear to me what are the functional requirements for this worker as I do not have  clear use cases for it. For example, I do not know how to answer to questions like: How many functions? What rate? What minimum function execution time? Etc.

Rutgers has a worker prototype that we wrote as a feasibility study to execute Python functions, but I am not sure this is what you are looking for/need. If you think it would be useful, you would be welcome to have a look at the code and we would be happy to go through it together.

**Comment by andre-merzky** [comment]
> @andre-merzky, would you agree with that?

Yes, and I also agree with your point about serialization and call context.  What function call context is required and how it is specified / provided will determine how we are going to implement the worker.  our prototype might be a good starting point, we'll see.

**Comment by andre-merzky** [comment]
I should add that we also may need changes to our client API and internal protocols, depending how mach and what kind of information we need to communicate for the call context.

**Comment by eirrgang** [comment]
> we have to support both function- and executable-based tasks

I think we are in agreement to this point.

By my comment, I meant that it is preferable to make command-line executables a subset of Python function use cases than the other way around, and it is the sufficient to focus on the Python function use case. If we are not in agreement here, then I can table that point for the moment. Alternatively, if there is an issue of software structure, then I guess I am proposing that we identify the layer and interface at which an operation implemented by a command line tool is substantially the same as an operation implemented with a Python interface, and use that to anchor the top-down aspects of the design.

> discussing how the function and its context should be passed from the user-facing 'client' to the worker

That sounds good. The serialized work graph was our proposal for conveying the client-requested work while allowing the actual executed tasks to be fully specified during execution. We have previously left the response unspecified, and I hope to clarify the runtime activity in issues such as #9 

gmxapi < 0.0.8 explored the serialization->deserialization->execution details more fully than gmxapi 0.0.8/0.1, but I can proceed with the next generation specification and prototyping in the context of this conversation.

> How many functions?

Functions implemented: < 10 in the core package (or, likely, in any single client/extension package); infinitely extensible

Functions executed: likely 1-30 unique functions executed in a single script; likely 1-5 unique functions executed in a single worker; likely 1-100 function-based tasks executed in a worker lifetime / execution session.

> What rate? What minimum function execution time?

Minimum function execution time could be microseconds, but we expect the most interesting work will be simulation tasks on the order of seconds to minutes, with less interesting work (from the standpoint of SCALE-MS value added opportunities) consisting of simulation operations on the order of hours to weeks. This is why we are interested in the overhead of data placement and task switching where we think we can reduce it from minutes to seconds or less.

A scenario we encounter regularly is an iteration sequence like 5-minute simulation -> 10-second data placement -> sub-second analysis -> 5-minute simulation that has been manually optimized down from several minutes of data placement and analysis, but which could be optimized to even less data placement cost with better API integration, which would in turn allow better statistics from a method by supporting more (shorter) iterations.

> What function call context is required and how it is specified

I think this has been the goal of a lot of the work so far on our end.

Let me try to break apart some of these discussion points into separate topics. I'll update soon.

**Comment by andre-merzky** [comment]
> By my comment, I meant that it is preferable to make command-line executables a subset of Python function use cases than the other way around, and it is the sufficient to focus on the Python function use case.

Functions and Executables may not differ semantically, and I can see why the upper layers may want to treat them on equal footing.  On the execution layer though, they are very different beasts and imply different implementations (and partially different architectures).  For example, the notion of an execution environment is very different: for functions this refers a call context (name spaces, available modules and entities), for an executables it refers to things like shell environment, I/O redirection, etc.   Some of those may overlap (in principle a function call may look at `os.environment` and read from `STDIN`, but it's not common), for others its strictly different.  The main difference if process creation and placement: for function calls, we need to place a custom worker (our code) on the target core(s) which hosts the function call and ensures the right call context; for executables we do not need workers on the target nodes at all, but ensure process creation and execution context via system utilities.  Isolation is another topic we need to consider: executables are usually well isolated - consecutive or concurrent python functions running in the same worker are usually not.

For those reasons we would like to keep a somewhat separate discussion of both concepts, at least wrt. requirements and architecture / implementation details.  How that distinction is rendered on the API level (if at all) is a different question.

**Comment by andre-merzky** [comment]
> Minimum function execution time could [anything]

I am rephrasing above, obviously :-)  but that's what we expected mostly.  The overhead of running microsecond functions on a remote node is likely significant, no matter what we do, and we would like to communicate that clearly to manage expectations to the execution layer.  There is only so much we can do wrt. latency hiding of the communication required to get the function to the node and back again...  In general, the expected execution time will put a limit to the efficiency at which we can execute at scale.  We feel fairly comfortable with everything running longer than a couple of minutes, at scale.

**Comment by eirrgang** [comment]
> In general, the expected execution time will put a limit to the efficiency at which we can execute at scale. We feel fairly comfortable with everything running longer than a couple of minutes, at scale.

That is our expectation, too, so we are trying to describe cases where we think we can turn groups of short tasks into longer work loads, and to identify the types of resource persistence or software interactions necessary. I think the next step for me, then, is to flesh out #7 

> ... for executables ...

Thank you for clarifying the distinctions. I'm not sure how the use cases handled by such a launch mechanism interact with any use case I've considered previously, so I'll have to chew on that...
